Add cover image png to serialization of magazine.

SashaLoayza opened the pull request:
<!-- IF A SECTION IS NOT APPLICABLE TO YOU, PLEASE DELETE IT!! -->

<!-- Your title should be able to summarize what changes you've made in one sentence. For example: "Exclude staff from the check for follows". For stacked PRs, please indicate clearly in the title where in the stack you are. For example: "[Eatery Refactor][4/5] Converted all files to MVP model" -->


## Overview

I changed the output of download_magazine in src/magazine.py to be a tuple of the appdev-upload services response objects for both the magazine pdf and the cover image as a png. Previously, it only returned the pdf response.

To accommodate this change, I changed the Magazine's serialize method to include the imageURL field in the outputted python dictionary.

Two new libraries were required to make this change, PyPdfium2 (a pdf library) and Pillow (an image library).



## Changes Made

The changes made were to two methods of the Magazine class in magazine.py: <download_magazine> and <serialize>.

In <download_magazine>, the pdf is loaded as before into a BytesIO buffer, and then it is loaded as a PdfDocument object of PyPdfium2. The first page is retrieved, converted to a png using PIL, and scaled to 150x220 pixels (all within a BytesIO buffer) and then uploaded to the Digital Ocean appdev-upload service. The tuple containing the two response objects from the upload service is returned (and subsequently used by the <serialize> method.

In <serialize>, the tuple is parsed into the pdf url and image url, and the python dictionary that is returned was modified to include the png url. The method checks that both requests were successful.  


## Test Coverage

I tested this feature manually by running the micro service on many magazines and checking that the png's were properly uploaded to Digital Ocean, and that the new imageURL field was added.


## Next Steps (delete if not applicable)

Need to complete the migrations on dev and prod server (add the imageURL field by reprising the magazines). The issue of large file sizes is still a problem. 


## Related PRs or Issues (delete if not applicable)
73 in volume-backend (add the imageURL field to the mongo and graphql schemas.

## Screenshots (delete if not applicable)

<img width="1249" alt="Screen Shot 2023-04-03 at 9 14 14 PM" src="https://user-images.githubusercontent.com/104698418/229666448-596a3419-b8cd-42f3-89a5-8acf5045c124.png">

  <summary>Microservice running and resulting png files</summary>

<img width="1440" alt="Screen Shot 2023-04-03 at 9 35 58 PM" src="https://user-images.githubusercontent.com/104698418/229666701-f959d95a-2cbf-4a78-97bf-fb1f5b6155cb.png">


  <summary>Magazine with imageURL field added to db (required new volume-backend pr 73 to run)</summary>

SashaLoayza committed:
Add cover image png to serialization of magazine.

snajima commented on src/magazine.py:
```diff
@@ -40,39 +47,61 @@ def download_magazine(self, id):
+                "image": "data:application/pdf;base64,"
+                + base64.b64encode(file.getvalue()).decode("utf-8")})
+            
+            pdf_response  = requests.post("https://upload.cornellappdev.com/upload/",pdf_payload).text
```
delete one of the two spaces before the `=`

snajima commented on src/magazine.py:
```diff
@@ -40,39 +47,61 @@ def download_magazine(self, id):
+            imageBytes = io.BytesIO() #create bytes stream to hold image
+            pil_image.save(imageBytes, format='png')
+            
+            image_payload = json.dumps(
+            {
+            "bucket": "volume",
+            "image": "data:image/png;base64,"
+            + base64.b64encode(imageBytes.getvalue()).decode("utf-8")})
+            
+            image_response = requests.post("https://upload.cornellappdev.com/upload/",image_payload).text
+
+            pdf_payload = json.dumps(
+                {"bucket": UPLOAD_BUCKET,
+                "image": "data:application/pdf;base64,"
+                + base64.b64encode(file.getvalue()).decode("utf-8")})
```
let's make the way we present dictionaries consistent. i think the best way is

```
json.dumps({
    "bucket": "volume",
    "image": "data:image/png;base64,"
    + base64.b64encode(imageBytes.getvalue()).decode("utf-8")
})
```
or
```
json.dumps(
    {
        "bucket": "volume",
        "image": "data:image/png;base64,"
        + base64.b64encode(imageBytes.getvalue()).decode("utf-8")
    }
)
```

snajima commented on src/magazine.py:
```diff
@@ -26,9 +27,15 @@ def __init__(self, sheet_row, publication):
         creds = service_account.Credentials.from_service_account_file(
-            os.getenv("GOOGLE_SERVICE_ACCOUNT_PATH")
-        )
+            os.getenv("GOOGLE_SERVICE_ACCOUNT_PATH"))
```
i think the parenthesis belong in the same indentation as the `creds = ...` line so this line's change should be reverted

snajima reviewed (approved):
good job! a few style nits so pls address them before merging :)

SashaLoayza committed:
Fix style in magazine.py

isaachan100 commented on publications.json:
```diff
@@ -421,7 +397,7 @@
       "contentTypes": ["articles"],
       "rssName": "The Cornell Daily Sun",
-      "rssURL": "https://cornellsun.com/feed/",
+      "rssURL": null,
```
should this be null or not? I think you added the rssURL link back in a later PR

SashaLoayza replied:
yes you are correct, I will keep the rssurl

isaachan100 commented on src/magazine.py:
```diff
@@ -40,39 +48,62 @@ def download_magazine(self, id):
-                "https://upload.cornellappdev.com/upload/",
-                payload,
-            ).text
+            pdf = pdfium.PdfDocument(file)
+            first_page = pdf[0]
+            pil_image = first_page.render().to_pil().resize((150,220))
+
+            imageBytes = io.BytesIO() #create bytes stream to hold image
+            pil_image.save(imageBytes, format='png')
+            
+            image_payload = json.dumps({
+                "bucket": "volume",
+                "image": "data:image/png;base64,"
+                + base64.b64encode(imageBytes.getvalue()).decode("utf-8")
+            })
```
in #25 i factored out download magazine into a utils.py so that we could use it for flyers too. would there be any way for you to make downloading the first page of the pdf a separate function or should i make changes to my pr?

SashaLoayza replied:
What would be best for me is if you slightly changed your refactoring by splitting up download_content into two functions: One that returns the Bytes_IO object of the content, and another that takes in a Bytes_IO and posts it to digital ocean, returning the same thing that download_magazine currently does. 

For this second function, it could either be specific to the content-type (pdf, image) or could take in a certain content type as a parameter. Attached is an updated utils.py with how you might do it with separate content-types for pdf and image.

[utils_new.txt](https://github.com/cuappdev/volume-microservice/files/11314435/utils_new.txt)

SashaLoayza replied:
Note that with this approach you will have to import pdfium2 as pdfium in the file as I did in utils_new.txt- This will also require adding pdfium2 and adding:
Pillow==9.5.0
pypdfium2==4.4.0

to the requirements.txt file.

isaachan100 reviewed (changes requested):
just some small comments and maybe a structural issue

SashaLoayza committed:
Keep cornell sun RSS feed url

As Isaac commented, the RSS feed url is included in a later commit as the issue is resolved.

SashaLoayza committed:
Refactor magazine.py to use utils.py

isaachan100 reviewed (approved):
lgtm!

SashaLoayza committed:
Merge branch 'main' into sasha/mag-first-page
